### ISSUE TITLE
Is gem still relevant with Rails 7.1?

### ISSUE THREAD

**Issue by fragkakis** [composite-primary-keys/composite_primary_keys#608]
Rails 7.1 (currently available as RC) supports [composite keys](https://edgeguides.rubyonrails.org/7_1_release_notes.html#composite-primary-keys).

Does it make sense to use this gem in Rails 7.1 or above? Is the Rails support equivalent, or does the gem offer anything extra?

**Comment by codeodor** [comment]
I probably would not consider using CPK for a new project, if you're just now starting one.

While I haven't checked out Rails 7.1 yet, for my project, if it's easy enough to migrate over, and isn't hampered by performance issues, we'll be doing that. I only plan on keeping CPK updated in our project if it's not easy enough to switch to Rails' version. (Note I am not the decision maker on this -- just giving my 2 cents, having contributed some updates to this gem)

**Comment by tomhughes** [comment]
As far as I can tell that comment in the release notes is at best misleading - if you read the linked blog post about what Shopify actually did you'll see that while they had a composite primary key in the database they still had a unique ID field which they used as the "primary key" in rails.

I've seen no evidence so far that rails 7.1.0 can actually handle multipart primary keys as a way of identifying records or joining tables - if it is possible then there is no documentation that I can find on how to do it.

**Comment by tomhughes** [comment]
I take it all back - the documentation is at https://guides.rubyonrails.org/active_record_composite_primary_keys.html and it does seem to work.

The bizarre things is that while you can set `primary_key` as an array, or let rails determine that automatically, you can't set `:foreign_key` to an array on a relationship and you have to use `:query_constraints` instead.

**Comment by jarl-dk** [comment]
I agree with you, @tomhughes, the need for `query_constraints` seem very odd. Like CPK is not a first class citizen after all.
I have tried it out as well. The docs you are referring to illustrates an example of a parent-child association, where the parrent has a CPK. But that is not the most common situation IMHO. the more common situation is that we have a parent-child association where the child has a CPK that is a composition of the PK of the parent plus one row in the child model. That case does not work very well in Rails 7.1. I'll see if I can open an issue on the rails project regarding that.

**Comment by jarl-dk** [comment]
I would love to see this gem to support Rails 7.1 until the CPK implementation in Rails is mature.

**Comment by tomhughes** [comment]
I was able to successfully switch https://github.com/openstreetmap/openstreetmap-website to rails 7.1 with the builtin CPK support in https://github.com/openstreetmap/openstreetmap-website/pull/4278 basically just be either dropping the primary key assignment or changing `primary_keys` to `primary_key` and then replacing compound foreign keys with `query_constraints` in https://github.com/openstreetmap/openstreetmap-website/pull/4278/commits/a0fb8adc8bff7cdcf8a4547ce3d2ed8ff1a44bb4.

**Comment by cfis** [comment]
Thanks for the info @tomhughes. So it seems there is no point in doing a 7.1 release for CPK then?

**Comment by jarl-dk** [comment]
FYI: I have opened a bug report on rails regarding composite primary keys:
https://github.com/rails/rails/issues/49597

**Comment by istrasci** [comment]
> Thanks for the info @tomhughes. So it seems there is no point in doing a 7.1 release for CPK then?

For whatever my vote is worth, I'd still like a 7.1 release and to use this gem for the foreseeable future.  I have my own gem that maintains compatibility with several Rails versions.  I'd rather not have to put logic in my gem's models just to use different syntax depending on the Rails version.  (I also don't want to immediately drop support for all Rails < 7.1.)

**Comment by gstokkink** [comment]
I'd also like a release for Rails 7.1. Best to upgrade in small steps I think, and dropping both this gem and upgrading to Rails 7.1 in one go is too much.

**Comment by jarl-dk** [comment]
Based on https://github.com/rails/rails/issues/49622#issuecomment-1766459466, I also also suggest maintaining this gem until Rails 8.

**Comment by Little-Rubyist** [comment]
Hello!
We are using CPK and attempting to upgrade to Rails 7.1.
However, due to the current limitation that `ibmdb` only supports up to Rails 7.0, it seems challenging to upgrade straightforwardly.

I'm curious to know if there are plans for CPK to support Rails 7.1.

**Comment by cfis** [comment]
The comments above indicate that switching to Rails built-in support for composite keys is not hard to do. In addition, upgrading CPK for major new Rails versions is quite difficult and time consuming (at least days of work if not a week or two). And I suspect this upgrade might be quite difficult since Rails 7.1 has likely changed most of the code CPK changes.  

Between those two facts, I hope it will *not* be necessary to upgrade this gem for 7.1. So I don't plan on updating CPK for Rails 7.1. I am happy to review PRs if someone wants to try though.

**Comment by jarl-dk** [comment]
Yet another reason for maintaining this gem (until Rails has matured on CPK): https://github.com/rails/rails/issues/51231

**Comment by jarl-dk** [comment]
FYI: Still bugs in rails:
https://github.com/rails/rails/issues/54881
https://github.com/rails/rails/issues/54882
